### ISSUE TITLE
training text classifier with soft probabilistic labels

### ISSUE THREAD

**Issue by vmando** [amaiya/ktrain#125]
Hi, 

I'm trying to train a text binary classifier with soft probabilistic training labels. In other words, for every training sample the training label is a 2-dimensional vector whose 2 entries are probabilities of belonging to negative and positive class respectively. Is there a way to train e.g. BERT classifier with these soft labels? 

If I understand correctly, the `categorical_crossentropy` loss function with softmax activation [is used](https://github.com/amaiya/ktrain/blob/9423c8bb908392f7d93b05b0656cd8e1e80a4856/ktrain/text/models.py#L130) for binary classification, and should be able to handle soft labels with no changes required. But I am not sure about the preprocessing, since that seems to require that training samples are classified into positive and negative examples. For instance the `text_from_folder` function in the [tutorial](https://nbviewer.jupyter.org/github/amaiya/ktrain/blob/master/tutorials/tutorial-04-text-classification.ipynb)  requires positive and negative examples to be in different subfolders, and also the other text loading methods like `text_from_df` seem to require hard labels.

Thanks!

**Comment by amaiya** [comment]
Hello,

This isn't an intended use case, but, in theory, it might still work.

I haven't tested it, but assuming you have a pandas dataframe that looks something like this:
```
 	text 	                neg 	pos
0 	The book is bad. 	0.95 	0.05
1 	This book is bad. 	0.95 	0.05
2 	The book is good. 	0.05 	0.95
3 	This book is good. 	0.05 	0.95
...
```
You can try something like this:
```python

import ktrain
from ktrain import text
MODEL_NAME = 'distilbert-base-uncased'
labels = ['neg', 'pos']
t = text.Transformer(MODEL_NAME, maxlen=25, class_names=labels)
trn = t.preprocess_train(df.text.values, df[labels].values)
model = t.get_classifier()
learner = ktrain.get_learner(model, train_data=trn)
```
or you could feed `df.text.values` and `df[labels].values` to `texts_from_array` also.

**Comment by amaiya** [comment]
@vmando:  From another issue, I realized that the `transformers` library (and, as a result, ktrain) cast labels to integers for classification problems.  So, this will make all of your softmax probability labels zero, which is problematic.  For regression problems in TensorFlow models, the `transformer` library seems to also (possibly incorrectly) cast dependent variable to int64.  I am looking into this and will try to correct this so float values are handled correctly wen using transformers for regression problems.

In the meantime, a possible workaround for you might be to transform your labels to integer values (eg., 0-1000) and treat problem as regression problem.

**Comment by vmando** [comment]
Thank you @amaiya . I did do the training using text_from_df and pandas dataframe as in your initial example, and it did seem to work and I got meaningful results. Could it be that Bert model specifically doesn't use the `transformers` library as the other models?

**Comment by amaiya** [comment]
@vmando You're exactly right.  When you use `bert` as argument to `text_classifier` function, *ktrain* currently uses an implementation of BERT from a different library, which allows you to avoid the current `transformers` issue.   Glad to hear it is working.
